### PR TITLE
8287925: AArch64: intrinsics for compareUnsigned method in Integer and Long

### PR DESCRIPTION
**Pull request by shqking**
x86 implemented the intrinsics for compareUnsigned() method in Integer and Long. See JDK-8283726. We add the corresponding AArch64 backend support in this patch.

Note-1: minor style issues are fixed for CmpL3 related rules.

Note-2: Jtreg case TestCompareUnsigned.java is updated to cover the matching rules for "comparing reg with imm" case.

Testing: tier1~3 passed on Linux/AArch64 platform with no new failures.

Following is the performance data for the JMH case:

```
                                                       Before          After
Benchmark                         (size) Mode  Cnt   Score   Error  Score   Error  Units
Integers.compareUnsignedDirect      500  avgt    5   0.994 ± 0.001  0.872 ± 0.015  us/op
Integers.compareUnsignedIndirect    500  avgt    5   0.991 ± 0.001  0.833 ± 0.055  us/op
Longs.compareUnsignedDirect         500  avgt    5   1.052 ± 0.001  0.974 ± 0.057  us/op
Longs.compareUnsignedIndirect       500  avgt    5   1.053 ± 0.001  0.916 ± 0.038  us/op
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287925](https://bugs.openjdk.org/browse/JDK-8287925): AArch64: intrinsics for compareUnsigned method in Integer and Long

### Reviewers
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to [ef39db22](https://git.openjdk.org/jdk/pull/11383/files/ef39db220c022412d01fdf10ac8c883187f54fe1)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11383/head:pull/11383` \
`$ git checkout pull/11383`

Update a local copy of the PR: \
`$ git checkout pull/11383` \
`$ git pull https://git.openjdk.org/jdk pull/11383/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11383`

View PR using the GUI difftool: \
`$ git pr show -t 11383`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11383.diff">https://git.openjdk.org/jdk/pull/11383.diff</a>

</details>

### CODE REVIEW

**Review by adinn — APPROVED** [review]
Looks good.

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -9645,20 +9645,113 @@ instruct get_and_addIi_no_resAcq(indirect mem, Universe dummy, immIAddSub incr)
...
+      "csetw $dst, ne\n\t"
+      "cnegw $dst, lo\t# CmpUL3(imm)"
+  %}
+  ins_encode %{
+    int32_t con = (int32_t)$src2$$constant;
+    if (con < 0) {
+      __ adds(zr, $src1$$Register, -con);
+    } else {
+      __ subs(zr, $src1$$Register, con);
+    }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:9734]:
```suggestion
    __ subs(zr, $src1$$Register, (int32_t)$src2$$constant);
```
should work here.

**Comment by theRealAph** [comment]
It seems to me that the enhancement here, if it exists, is in the noise. That may well be because the test is dominated by memory traffic.

**Comment by theRealAph** [comment]
I've been trying for an hour or so to write a benchmark that is actually improved by this patch. I have been unable to do so. Part of the problem is that using `cmp; cset; cneg` doesn't take advantage of branch prediction. And more often than not, the result of a comparison is predictable. Also, the code isn't necessarily shorter either.
I do not believe this patch should be committed unless someone manages to write a benchmark that demonstrates some advantage.

**Comment by theRealAph** [comment]
What would probably work better is to idealize `(cmp (cmp3 x y) < 0)` to `(cmpU x y)`

**Comment by shqking** [comment]
> What would probably work better is to idealize `(cmp (cmp3 x y) < 0)` to `(cmpU x y)`

I think this has been done in the original x86 patch. See https://github.com/openjdk/jdk/pull/9068/files#diff-054ecd9354722843f23556a38d2c24546c8a777b58b3442abea2d5e9fe6bb916R851

**Comment by theRealAph** [comment]
> > What would probably work better is to idealize `(cmp (cmp3 x y) < 0)` to `(cmpU x y)`
> 
> I think this has been done in the original x86 patch. See https://github.com/openjdk/jdk/pull/9068/files#diff-054ecd9354722843f23556a38d2c24546c8a777b58b3442abea2d5e9fe6bb916R851

Interesting. I didn't see this happening. I'll have another look.

**Comment by shqking** [comment]
> I've been trying for an hour or so to write a benchmark that is actually improved by this patch. I have been unable to do so. Part of the problem is that using `cmp; cset; cneg` doesn't take advantage of branch prediction. And more often than not, the result of a comparison is predictable. Also, the code isn't necessarily shorter either. I do not believe this patch should be committed unless someone manages to write a benchmark that demonstrates some advantage.

Thanks for your insight comment. I'm doing some investigation.

**Comment by shqking** [comment]
Hi @theRealAph ,

> It seems to me that the enhancement here, if it exists, is in the noise. That may well be because the test is dominated by memory traffic.

Yes. I agree.  We should use more `compareUnsigned` or use less memory loading in the loop.
I will show you the new test data later.

> Part of the problem is that using `cmp; cset; cneg` doesn't take advantage of branch prediction. And more often than not, the result of a comparison is predictable.

Yes. I agree.

I'd like to share my investigation here.

### Motivation of this intrinsic
I checked the two PRs in [JDK-8283726](https://bugs.openjdk.org/browse/JDK-8283726), i.e. the initial x86 intrinsic task.
From the discussions of https://github.com/openjdk/jdk/pull/7975 and https://github.com/openjdk/jdk/pull/9068, I think there are two motivations to introduce this `compareUnsigned` intrinsic.

**motivation-1**:  
> the compiler can recognise the pattern x + MIN_VALUE < y + MIN_VALUE and transforms it into x u< y. This transformation is fragile however if one of the arguments is in the form x + con

See the discussion [here](https://github.com/openjdk/jdk/pull/7975#issuecomment-1082464476).

I think that's why `bound - 16` is used rather than simply `bound` in [the JMH case](https://github.com/openjdk/jdk/pull/9068/files#diff-0a47e045a0f44094a7ec1fe7a251cc1d99d9fcd9c330cb33a79dd58e7f21b5d6R163).

**motivation-2**:
An optimization can be introduced, i.e. idealizing `(cmp (cmp3 x y) < 0) to (cmpU x y)`

There is `compareUnsignedIndirect` function in the JMH case to show the performance uplift of this optimization.

### Update the JMH case

I'd like to use `Integer` case as an illustrating example, and I think `Long` can be handled in the similar way.
Here shows my update to the JMH case.
```
    // Compare between random values.
    // Use big loops
    // Operate with another constant before cmpU3
    @Benchmark
    public void compareUnsignedDirect2(Blackhole bh) {
        int r = 0;
        int inx1 = 0;
        int inx2 = 0;
        int i, e1, e2;
        for (i = 0; i < size; i++) {
            inx1 = intsTiny[i];
            inx2 = intsSmall[i];
            for (i = 0; i < size; i++) {
                inx1 = (inx1 + i) % size;
                inx2 = (inx2 + i) % size;
                e1 = intsBig[inx1];
                e2 = intsBig[inx2];
                r += Integer.compareUnsigned(e1, e2 - 16);
                r -= Integer.compareUnsigned(r, e1 - 16);
                r += Integer.compareUnsigned(r, e2 - 16);
            }
        }
        bh.consume(r);
    }

    @Benchmark
    public void compareUnsignedIndirect2(Blackhole bh) {
        int r = 0;
        int inx1 = 0;
        int inx2 = 0;
        int i, e1, e2;
        for (i = 0; i < size; i++) {
            inx1 = intsTiny[i];
            inx2 = intsSmall[i];
            for (i = 0; i < size; i++) {
                inx1 = (inx1 + i) % size;
                inx2 = (inx2 + i) % size;
                e1 = intsBig[inx1];
                e2 = intsBig[inx2];
                r += (Integer.compareUnsigned(e1, e2 - 16) < 0) ? 1 : 0;
                r -= (Integer.compareUnsigned(r, e1 - 16) < 0) ? 1 : 0;
                r += (Integer.compareUnsigned(r, e2 - 16) < 0) ? 1 : 0;
            }
        }
        bh.consume(r);
    }
```
Note-1: random values are passed to `compareUnisnged` to evaluate `branch predication` affect. Inevitably, memory load is introduced in the inner loop, i.e. getting the values of `e1` and `e2`.
Note-2: big loops are used. In my evaluation, I passed `10240` to parameter `size`.
Note-3: more `compareUnisnged` operations are done in the inner loop.
Note-4: similarly, we pass `- 16` to `compareUnsigned` and use `Direct/Indirect` versions.

### Evaluation data on x86 and aarch64

Note-1: `before` means `disable intrinsic _compareUnsigned_i`, and `after` means `enable it`.
Note-2: the unit is `us/op`. The smaller, the better.

For case **compareUnsignedDirect2**

```
on AArch64:
  after : 55.144 ± 8.887  us/op
  before: 51.423 ± 6.891  us/op

on x86:
  after : 70.549 ± 3.729  us/op
  before: 67.781 ± 2.876  us/op
```

For case **compareUnsignedIndirect2**
```
on AArch64:
  after : 48.915 ± 11.054  us/op
  before: 52.782 ±  8.264  us/op

on x86:
  after : 68.107 ± 3.768  us/op
  before: 70.231 ± 3.966  us/op
```

From the evaluation data, we can see
1) There are **performance regression** for case **compareUnsignedDirect2** on both aarch64 and x86.
I checked the C2 generated code for this case on aarch64, and I don't think `e - 16` can prevent the C2 transformation " x+min_val OP y+min_val -> x uOP y". 
I think C2 can generate good enough code and our intrinsic didn't win.
Perhaps, **motivation-1** is not accurate??

2) There are **slight performance uplifts** for case **compareUnsignedIndirect2** on both aarch64 and x86.
Hence, the idealization optimization (See **motivation-2**) works.
I checked the generated code with this optimization and the sequence `cmp; cset; cneg` is gone.

With the investigation above, I personally don't have strong reason to introduce this intrinsic to aarch64 part.
WDYT? @theRealAph 

Besides, it would be nice if you could take a look at this discussion @merykitty?
Is there anything I missed or misunderstood? Thanks in advance.

**Comment by theRealAph** [comment]
Try this one:

```
    @Benchmark
    public int compareUnsignedDirect(Blackhole bh) {
        int probe1 = seed, probe2 = seed ^ seed << 5;
        int sum = 0;
        for (int i = 0; i < size; i++) {
            probe1 ^= probe1 << 13;
            sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 << 13;
            probe1 ^= probe1 >>> 17;                            sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
            sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 >>> 17;
            probe1 ^= probe1 << 5;                              sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
            sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 << 5;
                                                                sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
        }
        seed = probe2 + probe1;
        return sum;
    }

```

Does that help?

**Comment by merykitty** [comment]
The motivation for these intrinsics, aside from unsigned comparison being a fairly basic operation, is for range checks of a load/store, which has the form of `0 <= offset && offset <= length - size`, by transforming this into `0 <= length - size && offset u<= length - size`, the first comparison as well as the computation of `length - size` can be hoisted out of the loop, which results in a single operation being loop varying. The reason this may not be recognised effectively by the idealiser is that if `size` is a constant, `(length - size) + MIN_VALUE` being folded into `length + (MIN_VALUE - size)`, breaks the pattern.

If we simply want to throw an exception in out-of-bound cases, then `Precondition::checkIndex` may suffice. This however may not be adequate if:

- We want to do something else. If the hardware does not support masked load, currently we do a load followed by a blend if the whole vector is inbound and fall back out of intrinsic otherwise.
- The bound is not provably loop-invariant, and not obviously non-negative. This may arise in `ArrayList` accesses, where bound checks are performed against the `size` field, which may need to be reloaded on each iteration and not obviously nonnegative to the compiler.

IMO the direct result of the method is less important, because the contract does not have any promise with respect to the exact return value, and the only thing that can be done with it is to compare it with 0, which will certainly be folded into a `CmpU` node.

@shqking I think your benchmark is not good, as the error is higher than the actual difference between samples, and the cost of the division may swallow everything inside the loop.

Thanks a lot.

**Comment by shqking** [comment]
> Try this one:
> 
> ```
>     @Benchmark
>     public int compareUnsignedDirect(Blackhole bh) {
>         int probe1 = seed, probe2 = seed ^ seed << 5;
>         int sum = 0;
>         for (int i = 0; i < size; i++) {
>             probe1 ^= probe1 << 13;
>             sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 << 13;
>             probe1 ^= probe1 >>> 17;                            sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
>             sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 >>> 17;
>             probe1 ^= probe1 << 5;                              sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
>             sum += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;        probe2 ^= probe2 << 5;
>                                                                 sum += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) /* <= 0 ? 1 : 0 */;
>         }
>         seed = probe2 + probe1;
>         return sum;
>     }
> ```
> 
> Does that help?

Thanks for your comment.

Regarding your example, I made minor updates, shown as below:
```
diff --git a/test/micro/org/openjdk/bench/java/lang/Integers.java b/test/micro/org/openjdk/bench/java/lang/Integers.java
index 43ceb5d18d2..5ecbee26cab 100644
--- a/test/micro/org/openjdk/bench/java/lang/Integers.java
+++ b/test/micro/org/openjdk/bench/java/lang/Integers.java
@@ -167,6 +167,52 @@ public class Integers {
         }
     }

+    @Benchmark
+    public void compareUnsignedIndirect3(Blackhole bh) {
+        int seed = intsBig[0];
+        int probe1 = seed;
+        int probe2 = seed ^ seed << 5;
+        int r = 0;
+        for (int i = 0; i < size; i++) {
+            probe1 ^= probe1 << 13;
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+            probe2 ^= probe2 << 13;
+            probe1 ^= probe1 >>> 17;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+            probe2 ^= probe2 >>> 17;
+            probe1 ^= probe1 << 5;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+            probe2 ^= probe2 << 5;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE) <= 0 ? 1 : 0;
+        }
+        bh.consume(r);
+    }
+
+    @Benchmark
+    public void compareUnsignedDirect3(Blackhole bh) {
+        int seed = intsBig[0];
+        int probe1 = seed;
+        int probe2 = seed ^ seed << 5;
+        int r = 0;
+        for (int i = 0; i < size; i++) {
+            probe1 ^= probe1 << 13;
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            probe2 ^= probe2 << 13;
+            probe1 ^= probe1 >>> 17;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            probe2 ^= probe2 >>> 17;
+            probe1 ^= probe1 << 5;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            probe2 ^= probe2 << 5;
+            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
+        }
+        bh.consume(r);
+    }
+
     @Benchmark
     public void reverseBytes() {
         for (int i = 0; i < size; i++) {
```

Here shows the performance data of `compareUnsignedDirect3` on aarch64.
```
Integer.MAX_VALUE
                            Before         After          Unit
compareUnsignedDirect3      2.986 ± 0.019  1.858 ± 0.008  us/op
```
We can see using this intrinsic can introduce some performance uplifts.

Regarding `Integer.compareUnsigned(XX, YY)`, we should use random values for both `XX` and `YY` in order to expose the advantage of branch predication.
Hence, I made the following two versions of updates.

**Version-2**: Using another constant value for `YY`, e.g., `-1`.
```
@@ -198,17 +198,17 @@ public class Integers {
         int r = 0;
         for (int i = 0; i < size; i++) {
             probe1 ^= probe1 << 13;
-            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe1, -1);
             probe2 ^= probe2 << 13;
             probe1 ^= probe1 >>> 17;
-            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
-            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe2, -1);
+            r += Integer.compareUnsigned(probe1, -1);
             probe2 ^= probe2 >>> 17;
             probe1 ^= probe1 << 5;
-            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
-            r += Integer.compareUnsigned(probe1, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe2, -1);
+            r += Integer.compareUnsigned(probe1, -1);
             probe2 ^= probe2 << 5;
-            r += Integer.compareUnsigned(probe2, Integer.MAX_VALUE);
+            r += Integer.compareUnsigned(probe2, -1);
         }
         bh.consume(r);
     }
```

**Version-3**: Using some random value for `YY`, e.g., `probe1, probe2, r`.
```
@@ -198,17 +198,17 @@ public class Integers {
         int r = 0;
         for (int i = 0; i < size; i++) {
             probe1 ^= probe1 << 13;
-            r += Integer.compareUnsigned(probe1, -1);
+            r += Integer.compareUnsigned(probe1, r);
             probe2 ^= probe2 << 13;
             probe1 ^= probe1 >>> 17;
-            r += Integer.compareUnsigned(probe2, -1);
-            r += Integer.compareUnsigned(probe1, -1);
+            r += Integer.compareUnsigned(probe2, probe1);
+            r += Integer.compareUnsigned(probe1, r);
             probe2 ^= probe2 >>> 17;
             probe1 ^= probe1 << 5;
-            r += Integer.compareUnsigned(probe2, -1);
-            r += Integer.compareUnsigned(probe1, -1);
+            r += Integer.compareUnsigned(probe2, probe1);
+            r += Integer.compareUnsigned(probe1, r);
             probe2 ^= probe2 << 5;
-            r += Integer.compareUnsigned(probe2, -1);
+            r += Integer.compareUnsigned(probe2, probe1);
         }
         bh.consume(r);
     }
```

Here shows all the performance data on aarch64.
```
Integer.MAX_VALUE (version-1)
--------------------------------------------------------------------
                            Before         After          Unit
compareUnsignedDirect2      2.986 ± 0.019  1.858 ± 0.008  us/op

-1 (version-2)
--------------------------------------------------------------------
                            Before         After          Unit
compareUnsignedDirect2      1.384 ± 0.002  1.858 ± 0.002  us/op


probe1,probe2,r (version-3)
--------------------------------------------------------------------
                            Before         After          Unit
compareUnsignedDirect2      2.488 ± 0.076  2.517 ± 0.013  us/op
```

Note that for **version-1 and version-2**:
1) **After** column: the data is nearly the same if using the intrinsic (), as our `cmp; cset; cneg` sequence is predictable.
2) **Before** column: the data differs a lot. IMO, the advantage of branch prediction is most utilized for **version-2**, since `-1` will be treated as the largest unsigned value, and most comparisons can be predicated.

Note that for **version-3** (i.e. both `XX` and `YY` are random values):
1) **Before** column: the data may vary a bit for different runs. Sometimes it ls slightly bigger than **After** column, that is, using the intrinsic can introduce performance uplift. Sometimes, it's slightly smaller than **After** column, that is, using the intrinsic can introduce performance regression.

**Comment by shqking** [comment]
> The motivation for these intrinsics, aside from unsigned comparison being a fairly basic operation, is for range checks of a load/store, which has the form of `0 <= offset && offset <= length - size`, by transforming this into `0 <= length - size && offset u<= length - size`, the first comparison as well as the computation of `length - size` can be hoisted out of the loop, which results in a single operation being loop varying. The reason this may not be recognised effectively by the idealiser is that if `size` is a constant, `(length - size) + MIN_VALUE` being folded into `length + (MIN_VALUE - size)`, breaks the pattern.
> 
@merykitty Thanks for your explanation. 
But I'm afraid I didn't fully get your point.
I think `offset u<= length - size` will be matched with `CmpU` node, rather than `CmpU3` node, right?

> @shqking I think your benchmark is not good, as the error is higher than the actual difference between samples, and the cost of the division may swallow everything inside the loop.
>
How about the case as suggested by aph, i.e. shown in the previous comment?

Thanks~

**Comment by merykitty** [comment]
@shqking

> But I'm afraid I didn't fully get your point.
I think offset u<= length - size will be matched with CmpU node, rather than CmpU3 node, right?

`CmpU3` serves the same purpose as `CmpL3` or `CmpD3`, that is to be an intermediate node that gets elided immediately. For example, this piece of code:

    static int square(long x, long y) {
        return x > y ? 1 : 0;
    }

gets compiled into:

    static int square(long, long);
        0: lload_0
        1: lload_2
        2: lcmp
        3: ifle          10
        6: iconst_1
        7: goto          11
       10: iconst_0
       11: ireturn

`lcmp` is parsed into a `CmpL3`, so the comparison part is parsed as `CmpI (CmpL3 x y) 0`, which is then transformed into `CmpL x y`.

Similarly, `Integer.compareUnsigned(x, y) > 0` is parsed as `CmpI (CmpU3 x y) 0`, which is then transformed into `CmpU x y`.

> How about the case as suggested by aph, i.e. shown in the previous comment?

I think it is better but I still think measuring the performance of `CmpU3` node is quite irrelevent.

Thanks.

**Comment by theRealAph** [comment]
The problem as I see it: the intrinsic results in worse performance for the 3-way case if the result is highly predictable.
And, in many cases such as bounds checking, the result will surely be highly predictable.
Therefore, on average, using this intrinsic for the 3-way case may not improve things and could make them worse.

Having said all of that, I believe that directly using the 3-way result may be so rare that we don't need to care about it. All that we really should optimize is the `Integer.compareUnsigned(x,y) cmp 0`.  As long as that doesn't regress, I'm happy.

I won't be working for the next few days, so please don't wait for any more replies from me.

**Comment by shqking** [comment]
@merykitty Thanks a lot for your prompt explanation.

Here is my understanding of node `CmpU3`.
1) Generation: As an intrinsic, `Integer.compareUnsigned()` will be compiled into `CmpU3` node. The corresponding JMH test case is `compareUnsignedDirect()` function.
2) Optimization: Similar to `CmpL3`, `CmpF3` and `CmpD3` nodes, idealization optimization is conducted, transforming `CmpI (CmpU3 x y) 0` into `CmpU x y`. As you mentioned, in this case `CmpU3` node works as an intermediate node. The corresponding JMH test case is `compareUnsignedIndirect()` function.

Here list two quotes from your previous comments.
> IMO the direct result of the method is less important, because the contract does not have any promise with respect to the exact return value, and the only thing that can be done with it is to compare it with 0, which will certainly be folded into a CmpU node.
>
> I think it is better but I still think measuring the performance of CmpU3 node is quite irrelevent.
>
I guess your point is that it's irrelevant to evaluate the performance of `CmpU3` node mainly because **in most scenarios, the result of `Integer.compareUnsigned(x, y)` would be compared with 0**, and `CmpU3` will be optimized out finally.

If my understanding is correct, may I ask how can we draw this conclusion, from observation or statistical results of the usages of `Integer.compareUnsigned(x, y)` methods in exisiting popular Java application?

Besides, I believe you may understand what we (aph and I) are concerned about already, but I'd like to describe it again in case I didn't make myself understood clearly. Our discussion was actually the performance of `CmpU3` node, because in AArch64, `cmp; cset; cneg` sequence would be generated, which is predictable compared to the C2 generated code, where the advantage of branch prediction can be utilized. From the performance data shown previously, I'm afraid it's hard to say the intrinsification code is always better than C2 compiled one.

Thanks.

**Comment by shqking** [comment]
Thanks for your comment. @theRealAph 
I didn't see your comment until I sent out my reply to @merykitty 's comment. :)
> The problem as I see it: the intrinsic results in worse performance for the 3-way case if the result is highly predictable. And, in many cases such as bounds checking, the result will surely be highly predictable. Therefore, on average, using this intrinsic for the 3-way case may not improve things and could make them worse.
> 
Yes. It seems so.

> Having said all of that, I believe that directly using the 3-way result may be so rare that we don't need to care about it. All that we really should optimize is the `Integer.compareUnsigned(x,y) cmp 0`.
> 
Yes. I think it's also @merykitty 's point.

**Comment by shqking** [comment]
Would you mind taking another look at the latest commit as suggested by aph? Thanks. @adinn

**Review by adinn — APPROVED** [review]
New changes are fine

File: src/hotspot/cpu/aarch64/aarch64.ad (hunk 1 above)
Comment (shqking) [reply src/hotspot/cpu/aarch64/aarch64.ad]: Updated.

**Comment by shqking** [comment]
Could you help take another look at this issue when you have spare time? Thanks @theRealAph

**Review by theRealAph — APPROVED** [review]

**Comment by shqking** [comment]
Thanks a lot for your reviews @theRealAph and @adinn 
Thanks for your invaluable input for the discussion @merykitty 
/integrate

**Pull request closed without merging**